hal/armv7a/zynq7000/timer.c

```diff
@@ -17,34 +17,140 @@
 #include "spinlock.h"
 #include "interrupts.h"
 
+#define TTC_SRC_CLK_CPU_1x 111111115 /* Hz */
 
 struct {
-	intr_handler_t wakeuph;
-	intr_handler_t timerh;
-	spinlock_t lock;
+	volatile u32 *ttc;
+	intr_handler_t handler;
+	volatile time_t jiffies;
+
+	u32 ticksPerFreq;
+	spinlock_t sp;
 } timer_common;
 
 
+enum {
+	clk_ctrl = 0, clk_ctrl2, clk_ctrl3, cnt_ctrl, cnt_ctrl2, cnt_ctrl3, cnt_value, cnt_value2, cnt_value3, interval_val, interval_cnt2, interval_cnt3,
+	match0, match1_cnt2, match1_cnt3, match1, match2_cnt2, match2_cnt3, match2, match3_cnt2, match3_cnt3, isr, irq_reg2, irq_reg3, ier, irq_en2,
+	irq_en3, ev_ctrl_t1, ev_ctrl_t2, ev_ctrl_t3, ev_reg1, ev_reg2, ev_reg3
+};
+
+
+extern unsigned int _end;
+
+
+void timer_jiffiesAdd(time_t t)
+{
+	(void)t;
+}
+
 
-extern void _end(void);
+void timer_setAlarm(time_t us)
+{
+	(void)us;
+}
+
+
+static int _timer_irqHandler(unsigned int n, cpu_context_t *ctx, void *arg)
+{
+	(void)n;
+	(void)arg;
+	(void)ctx;
+
+	u32 st = *(timer_common.ttc + isr);
+
+	/* Interval IRQ */
+	if (st & 0x1)
+		timer_common.jiffies += timer_common.ticksPerFreq;
+
+	/* Clear irq status */
+	*(timer_common.ttc + isr) = st;
+
+	hal_cpuDataSyncBarrier();
+
+	return 0;
+}
 
 
-/* TODO */
 void hal_setWakeup(u32 when)
 {
 
 }
 
 
-/* TODO */
 time_t hal_getTimer(void)
 {
-	return 0;
+	spinlock_ctx_t sc;
+	time_t ret;
+
+	hal_spinlockSet(&timer_common.sp, &sc);
+	ret = timer_common.jiffies;
+	hal_spinlockClear(&timer_common.sp, &sc);
+
+	return ret;
+}
+
+
+static void timer_setPrescaler(u32 freq)
+{
+	u32 ticks;
+	u32 prescaler;
+
+	ticks = TTC_SRC_CLK_CPU_1x / freq;
+
+	prescaler = 0;
+	while ((ticks >= 0xffff) && (prescaler < 0x10)) {
+		prescaler++;
+		ticks /= 2;
+	}
+
+	if (prescaler) {
+		/* Enable and set prescaler */
+		prescaler--;
+		*(timer_common.ttc + clk_ctrl) |= (prescaler << 1);
+		*(timer_common.ttc + clk_ctrl) |= 0x1;
```

Comment:
So if these bits not being zero won't mess our operations
```suggestion
		*(timer_common.ttc + clk_ctrl) = (*(timer_common.ttc + clk_ctrl) & ~0x1f) | (prescaler << 1);
		*(timer_common.ttc + clk_ctrl) |= 0x1;
```

Reply:
Done